src/init.cpp

```diff
@@ -68,8 +68,8 @@ using namespace std;
 CWallet* pwalletMain = NULL;
 int nWalletBackups = 10;
 #endif
-bool fFeeEstimatesInitialized = false;
-bool fRestartRequested = false; // true: restart false: shutdown
+volatile bool fFeeEstimatesInitialized = false;
+volatile bool fRestartRequested = false; // true: restart false: shutdown
 
```

Comment:
Why do these need to be volatile? As far as I can tell the compiler should be able to see/predict all the use cases of `fFeeEstimatesInitialized` and `fRestartRequired`. However, since `fRequestShutdown` and `fReopenDebugLog` handle the system interrupt signals (ln. 271, 276) the compiler needs to hold back from optimization because of the unpredictability of users sending system interrupts.

Am I missing something?

Reply:
As far as I know, C/C++ compiler would optimize out any variable (or function that checks the variable) that is modified only as a result of a signal (like a user requesting quit). Ran into a similar situation writing network code for a chat app. I can write a small demo application to demonstrate if we'd like to investigate further.

Reply:
Please leave reference to where `fFeeEstimates` and `fRestartRequired` are modified by a system signal? Also.... From what I remember, in C++11 standards `volatile` should only be used for hardware communication, where as `atomic` should be used for multithreading. Real solution to this would be to do the computation in its own thread, and not on the main thread, thus eliminating the need for a shutdown check during each iteration over the accumulators.

Reply:
RE: `fFeeEstimates` + `fRestartRequired` modified by system signal

Fair question, and admittedly there is an assumption on my part that these variables were modified by something like a system signal. 

According to Qt docs, the origin of any modification comes from the Qt windowing system ([ref](http://doc.qt.io/qt-4.8/qwidget.html#closeEvent)). I did not dig any deeper to figure out *for sure* if a compiler would treat such a scenario identically, but the behavior I observed in testing the associated bug is consistent with the compiler dropping the if statement completely due to aggressive optimizations. Inserting `volitile` keyword eliminated the observed behavior.

RE: `atomic` vs `volitile`

`atomic` is appropriate for situations in which many threads must share the same variable and race conditions arise, it ensures that access patterns in hardware do not cause trouble (ex. interleaved read-writes). As far as I can tell `fFeeEstimates` + `fRestartRequired` are not the subject of race conditions. Their modification is just unexpected to the compiler (I assume), hence `volitile` is more appropriate. Smarter people than me talk about this [here](https://stackoverflow.com/questions/8819095/concurrency-atomic-and-volatile-in-c11-memory-model).

RE: Real solution

Yes, I would love to offload much of the initialization logic to it's own thread. Just figured I'd deliver something that works, then spend more time on a more general solution for the project in the coming weeks.

Reply:
@rejectedpromise we don't use Qt in the business logic.

@aaronlangford31 most of the init procress needs to be done it a specific order. Probably there is not a case where splitting it off into a worker thread would provide any substantial benefit.

utACK, but needs to be squashed down before merge. Will test it soon.